lib/bamboo/adapters/mailgun_adapter.ex

```diff
@@ -0,0 +1,125 @@
+defmodule Bamboo.MailgunAdapter do
+  @moduledoc """
+  Sends email using Mailgun's API.
+
+  Use this adapter to send emails through Mailgun's API. Requires that an API
+  key and a domain are set in the config.
+
+  ## Example config
+
+      # In config/config.exs, or config.prod.exs, etc.
+      config :my_app, MyApp.Mailer,
+        adapter: Bamboo.MailgunAdapter,
+        api_key: "my_api_key",
+        domain: "your.domain"
+
+      # Define a Mailer. Maybe in lib/my_app/mailer.ex
+      defmodule MyApp.Mailer do
+        use Bamboo.Mailer, otp_app: :my_app
+      end
+  """
+
+  @base_uri "https://api.mailgun.net/v3/"
+  @behaviour Bamboo.Adapter
+
+  alias Bamboo.Email
+
+  defmodule ApiError do
+    defexception [:message]
+
+    def exception(%{params: params, response: response}) do
+      message = """
+      There was a problem sending the email through the Mailgun API.
+
+      Here is the response:
+
+      #{inspect response, limit: :infinity}
+
+
+      Here are the params we sent:
+
+      #{inspect params, limit: :infinity}
+      """
+      %ApiError{message: message}
+    end
+  end
+
+  def deliver(email, config) do
+    body = email |> to_mailgun_body |> Plug.Conn.Query.encode
+
+    case HTTPoison.post!(full_uri(config), body, headers(config)) do
+      %{status_code: status} = response when status > 299 ->
+        raise(ApiError, %{params: body, response: response})
+      response -> response
+    end
+  end
+
+  @doc false
+  def handle_config(config) do
+    for setting <- [:api_key, :domain] do
+      if config[setting] in [nil, ""] do
+        raise_missing_setting_error(config, setting)
+      end
+    end
+    config
+  end
+
+  defp raise_missing_setting_error(config, setting) do
+    raise ArgumentError, """
+    There was no #{setting} set for the Mailgun adapter.
+
+    * Here are the config options that were passed in:
+
+    #{inspect config}
+    """
+  end
+
+  defp headers(config) do
+    [
+      {"Content-Type", "application/x-www-form-urlencoded"},
+      {"Authorization", "Basic #{auth_token(config)}"},
+    ]
+  end
+
+  defp auth_token(config) do
+    Base.encode64("api:" <> config.api_key)
+  end
+
+  @mailgun_message_fields ~w(from to cc bcc subject text html)a
+
+  defp to_mailgun_body(%Email{} = email) do
+    email
+    |> Map.from_struct
```

Comment:
You *might* be able to remove this since the `Map.take` on line 96 will make it into a regular map